File: zig/tree234.zig
// Generic array-backed 2-3-4 tree with monoidal measures.
//
// Type parameters (via comptime Traits struct):
//   K      - key/element type (stored in sorted order)
//   M      - measure type (monoidal annotation on subtrees)
//   Traits - a struct providing:
//       fn measure(K) M              - measure a single element
//       fn combine(M, M) M           - monoidal combine
//       fn identity() M              - monoidal identity
//       fn compare(K, K) i32         - <0, 0, >0

const std = @import("std");
const Allocator = std.mem.Allocator;

pub fn Tree234(comptime K: type, comptime M: type, comptime Traits: type) type {
    return struct {
        const Self = @This();

        pub const Node = struct {
            n: u8, // number of keys: 1, 2, or 3
            keys: [3]K,
            children: [4]i32,
            measure: M,

            fn init() Node {
                return .{
                    .n = 0,
                    .keys = undefined,
                    .children = .{ -1, -1, -1, -1 },
                    .measure = Traits.identity(),
                };
            }
        };

        pub const WeightResult = struct {
            key: K,
            cum_before: f64,
            index: i32,
            found: bool,
        };

        nodes: std.ArrayListUnmanaged(Node),
        free_list: std.ArrayListUnmanaged(i32),
        root: i32,
        count: i32,
        allocator: Allocator,

        pub fn init(allocator: Allocator) Self {
            return .{
                .nodes = .empty,
                .free_list = .empty,
                .root = -1,
                .count = 0,
                .allocator = allocator,
            };
        }

        pub fn deinit(self: *Self) void {
            self.nodes.deinit(self.allocator);
            self.free_list.deinit(self.allocator);
        }

        fn allocNode(self: *Self) !i32 {
            if (self.free_list.items.len > 0) {
                const idx = self.free_list.items[self.free_list.items.len - 1];
                self.free_list.items.len -= 1;
                self.nodes.items[@intCast(@as(u32, @bitCast(idx)))] = Node.init();
                return idx;
            } else {
                const idx: i32 = @intCast(self.nodes.items.len);
                try self.nodes.append(self.allocator, Node.init());
                return idx;
            }
        }

        fn ix(self: *const Self, idx: i32) *Node {
            return &self.nodes.items[@intCast(@as(u32, @bitCast(idx)))];
        }

        fn ixConst(self: *const Self, idx: i32) *const Node {
            return &self.nodes.items[@intCast(@as(u32, @bitCast(idx)))];
        }

        fn isLeaf(self: *const Self, idx: i32) bool {
            return self.ixConst(idx).children[0] == -1;
        }

        fn is4Node(self: *const Self, idx: i32) bool {
            return self.ixConst(idx).n == 3;
        }

        fn recomputeMeasure(self: *Self, idx: i32) void {
            const nd = self.ixConst(idx);
            var m = Traits.identity();
            for (0..@as(usize, nd.n) + 1) |i| {
                if (nd.children[i] != -1) {
                    m = Traits.combine(m, self.ixConst(nd.children[i]).measure);
                }
                if (i < nd.n) {
                    m = Traits.combine(m, Traits.measure(nd.keys[i]));
                }
            }
            self.ix(idx).measure = m;
        }

        // Split a 4-node child at position child_pos of parent.
        fn splitChild(self: *Self, parent_idx: i32, child_pos: usize) !void {
            const child_idx = self.ixConst(parent_idx).children[child_pos];

            // Save child data before allocNode may invalidate pointers
            const k0 = self.ixConst(child_idx).keys[0];
            const k1 = self.ixConst(child_idx).keys[1];
            const k2 = self.ixConst(child_idx).keys[2];
            const c0 = self.ixConst(child_idx).children[0];
            const c1 = self.ixConst(child_idx).children[1];
            const c2 = self.ixConst(child_idx).children[2];
            const c3 = self.ixConst(child_idx).children[3];

            // Create right node with k2, c2, c3
            const right_idx = try self.allocNode();
            // After allocNode, re-access by index
            self.ix(right_idx).n = 1;
            self.ix(right_idx).keys[0] = k2;
            self.ix(right_idx).children[0] = c2;
            self.ix(right_idx).children[1] = c3;

            // Shrink child (left) to k0, c0, c1
            self.ix(child_idx).n = 1;
            self.ix(child_idx).keys[0] = k0;
            self.ix(child_idx).children[0] = c0;
            self.ix(child_idx).children[1] = c1;
            self.ix(child_idx).children[2] = -1;
            self.ix(child_idx).children[3] = -1;

            // Recompute measures for left and right
            self.recomputeMeasure(child_idx);
            self.recomputeMeasure(right_idx);

            // Insert mid_key (k1) into parent at child_pos
            const parent_n = self.ixConst(parent_idx).n;
            {
                var i: usize = parent_n;
                while (i > child_pos) : (i -= 1) {
                    self.ix(parent_idx).keys[i] = self.ixConst(parent_idx).keys[i - 1];
                    self.ix(parent_idx).children[i + 1] = self.ixConst(parent_idx).children[i];
                }
            }
            self.ix(parent_idx).keys[child_pos] = k1;
            self.ix(parent_idx).children[child_pos + 1] = right_idx;
            self.ix(parent_idx).n += 1;

            self.recomputeMeasure(parent_idx);
        }

        // Insert key into a non-full node's subtree.
        fn insertNonFull(self: *Self, idx: i32, key: K) !void {
            if (self.isLeaf(idx)) {
                // Insert key in sorted position
                var pos: usize = self.ixConst(idx).n;
                while (pos > 0 and Traits.compare(key, self.ixConst(idx).keys[pos - 1]) < 0) {
                    self.ix(idx).keys[pos] = self.ixConst(idx).keys[pos - 1];
                    pos -= 1;
                }
                self.ix(idx).keys[pos] = key;
                self.ix(idx).n += 1;
                self.recomputeMeasure(idx);
                return;
            }

            // Find child to descend into
            var pos: usize = 0;
            while (pos < self.ixConst(idx).n and Traits.compare(key, self.ixConst(idx).keys[pos]) >= 0) {
                pos += 1;
            }

            // If that child is a 4-node, split it first
            if (self.is4Node(self.ixConst(idx).children[pos])) {
                try self.splitChild(idx, pos);
                // After split, mid_key is at keys[pos]. Decide which side.
                if (Traits.compare(key, self.ixConst(idx).keys[pos]) >= 0) {
                    pos += 1;
                }
            }

            try self.insertNonFull(self.ixConst(idx).children[pos], key);
            self.recomputeMeasure(idx);
        }

        pub fn insert(self: *Self, key: K) !void {
            if (self.root == -1) {
                self.root = try self.allocNode();
                self.ix(self.root).n = 1;
                self.ix(self.root).keys[0] = key;
                self.recomputeMeasure(self.root);
                self.count += 1;
                return;
            }

            // If root is a 4-node, split it.
            if (self.is4Node(self.root)) {
                const old_root = self.root;
                self.root = try self.allocNode();
                self.ix(self.root).children[0] = old_root;
                try self.splitChild(self.root, 0);
            }

            try self.insertNonFull(self.root, key);
            self.count += 1;
        }

        pub fn clear(self: *Self) void {
            self.nodes.clearRetainingCapacity();
            self.free_list.clearRetainingCapacity();
            self.root = -1;
            self.count = 0;
        }

        pub fn size(self: *const Self) i32 {
            return self.count;
        }

        pub fn rootMeasure(self: *const Self) M {
            if (self.root == -1) return Traits.identity();
            return self.ixConst(self.root).measure;
        }

        // In-order traversal
        fn forEachImpl(self: *const Self, idx: i32, callback: anytype) void {
            if (idx == -1) return;
            const nd = self.ixConst(idx);
            for (0..@as(usize, nd.n) + 1) |i| {
                if (nd.children[i] != -1) {
                    self.forEachImpl(nd.children[i], callback);
                }
                if (i < nd.n) {
                    callback.call(nd.keys[i]);
                }
            }
        }

        // Collect all keys in-order into a slice
        pub fn collect(self: *const Self, out: *std.ArrayListUnmanaged(K), allocator: Allocator) !void {
            out.clearRetainingCapacity();
            try out.ensureTotalCapacity(allocator, @intCast(@as(u32, @bitCast(self.count))));
            self.collectImpl(self.root, out);
        }

        fn collectImpl(self: *const Self, idx: i32, out: *std.ArrayListUnmanaged(K)) void {
            if (idx == -1) return;
            const nd = self.ixConst(idx);
            for (0..@as(usize, nd.n) + 1) |i| {
                if (nd.children[i] != -1) {
                    self.collectImpl(nd.children[i], out);
                }
                if (i < nd.n) {
                    // We already ensured capacity, so this won't fail
                    out.appendAssumeCapacity(nd.keys[i]);
                }
            }
        }

        fn subtreeCount(self: *const Self, idx: i32) i32 {
            if (idx == -1) return 0;
            const nd = self.ixConst(idx);
            var c: i32 = @intCast(nd.n);
            for (0..@as(usize, nd.n) + 1) |i| {
                if (nd.children[i] != -1) {
                    c += self.subtreeCount(nd.children[i]);
                }
            }
            return c;
        }

        pub fn findByWeight(self: *const Self, target: f64, comptime weightOf: fn (M) f64) WeightResult {
            if (self.root == -1) return .{ .key = undefined, .cum_before = 0, .index = 0, .found = false };
            return self.findByWeightImpl(self.root, target, 0.0, 0, weightOf);
        }

        fn findByWeightImpl(self: *const Self, idx: i32, target: f64, cum: f64, global_idx: i32, comptime weightOf: fn (M) f64) WeightResult {
            if (idx == -1) return .{ .key = undefined, .cum_before = 0, .index = 0, .found = false };

            const nd = self.ixConst(idx);
            var running_cum = cum;
            var running_idx = global_idx;

            for (0..@as(usize, nd.n) + 1) |i| {
                // Process child
                if (nd.children[i] != -1) {
                    const child_weight = weightOf(self.ixConst(nd.children[i]).measure);
                    if (running_cum + child_weight >= target) {
                        return self.findByWeightImpl(nd.children[i], target, running_cum, running_idx, weightOf);
                    }
                    running_cum += child_weight;
                    running_idx += self.subtreeCount(nd.children[i]);
                }

                if (i < nd.n) {
                    const key_weight = weightOf(Traits.measure(nd.keys[i]));
                    if (running_cum + key_weight >= target) {
                        return .{ .key = nd.keys[i], .cum_before = running_cum, .index = running_idx, .found = true };
                    }
                    running_cum += key_weight;
                    running_idx += 1;
                }
            }

            return .{ .key = undefined, .cum_before = 0, .index = 0, .found = false };
        }

        pub fn buildFromSorted(self: *Self, sorted: []const K) !void {
            self.clear();
            if (sorted.len == 0) return;
            self.count = @intCast(sorted.len);
            self.root = try self.buildRecursive(sorted, 0, @intCast(sorted.len));
        }

        fn buildRecursive(self: *Self, sorted: []const K, lo: i32, hi: i32) !i32 {
            const n = hi - lo;
            if (n <= 0) return -1;

            if (n <= 3) {
                const idx = try self.allocNode();
                self.ix(idx).n = @intCast(@as(u32, @bitCast(n)));
                for (0..@as(usize, @intCast(@as(u32, @bitCast(n))))) |i| {
                    self.ix(idx).keys[i] = sorted[@intCast(@as(u32, @bitCast(lo + @as(i32, @intCast(i)))))];
                }
                self.recomputeMeasure(idx);
                return idx;
            }

            if (n <= 7) {
                const mid = lo + @divTrunc(n, 2);
                const left = try self.buildRecursive(sorted, lo, mid);
                const right = try self.buildRecursive(sorted, mid + 1, hi);
                const idx = try self.allocNode();
                self.ix(idx).n = 1;
                self.ix(idx).keys[0] = sorted[@intCast(@as(u32, @bitCast(mid)))];
                self.ix(idx).children[0] = left;
                self.ix(idx).children[1] = right;
                self.recomputeMeasure(idx);
                return idx;
            }

            // For larger ranges, use 3-node
            const third = @divTrunc(n, 3);
            const m1 = lo + third;
            const m2 = lo + 2 * third + 1;
            const c0 = try self.buildRecursive(sorted, lo, m1);
            const c1 = try self.buildRecursive(sorted, m1 + 1, m2);
            const c2 = try self.buildRecursive(sorted, m2 + 1, hi);
            const idx = try self.allocNode();
            self.ix(idx).n = 2;
            self.ix(idx).keys[0] = sorted[@intCast(@as(u32, @bitCast(m1)))];
            self.ix(idx).keys[1] = sorted[@intCast(@as(u32, @bitCast(m2)))];
            self.ix(idx).children[0] = c0;
            self.ix(idx).children[1] = c1;
            self.ix(idx).children[2] = c2;
            self.recomputeMeasure(idx);
            return idx;
        }
    };
}

File: kotlin/Tree234.kt
/**
 * Generic array-backed 2-3-4 tree with monoidal measures.
 *
 * Type parameters:
 *   K - key/element type (stored in sorted order)
 *   M - measure type (monoidal annotation on subtrees)
 *
 * The [traits] parameter provides:
 *   measure(K) -> M          - measure a single element
 *   combine(M, M) -> M       - monoidal combine
 *   identity() -> M           - monoidal identity
 *   compare(K, K) -> Int      - <0, 0, >0
 */

class Tree234<K, M>(private val traits: TreeTraits<K, M>) {

    interface TreeTraits<K, M> {
        fun measure(key: K): M
        fun combine(a: M, b: M): M
        fun identity(): M
        fun compare(a: K, b: K): Int
    }

    class Node<K, M>(identity: M) {
        var n: Int = 0                          // number of keys: 1, 2, or 3
        val keys: Array<Any?> = arrayOfNulls(3)
        val children: IntArray = intArrayOf(-1, -1, -1, -1)
        var measure: M = identity
    }

    data class WeightResult<K>(
        val key: K,
        val cumBefore: Double,
        val index: Int,
        val found: Boolean
    )

    private val nodes = ArrayList<Node<K, M>>()
    private val freeList = ArrayList<Int>()
    private var root: Int = -1
    private var count: Int = 0

    val size: Int get() = count

    fun rootMeasure(): M {
        if (root == -1) return traits.identity()
        return nodes[root].measure
    }

    private fun allocNode(): Int {
        return if (freeList.isNotEmpty()) {
            val idx = freeList.removeAt(freeList.size - 1)
            val nd = nodes[idx]
            nd.n = 0
            nd.keys[0] = null; nd.keys[1] = null; nd.keys[2] = null
            nd.children[0] = -1; nd.children[1] = -1; nd.children[2] = -1; nd.children[3] = -1
            nd.measure = traits.identity()
            idx
        } else {
            val idx = nodes.size
            nodes.add(Node(traits.identity()))
            idx
        }
    }

    private fun freeNode(idx: Int) {
        freeList.add(idx)
    }

    private fun isLeaf(idx: Int): Boolean = nodes[idx].children[0] == -1
    private fun is4Node(idx: Int): Boolean = nodes[idx].n == 3

    @Suppress("UNCHECKED_CAST")
    private fun nodeKey(idx: Int, pos: Int): K = nodes[idx].keys[pos] as K

    private fun recomputeMeasure(idx: Int) {
        val nd = nodes[idx]
        var m = traits.identity()
        for (i in 0..nd.n) {
            if (nd.children[i] != -1) {
                m = traits.combine(m, nodes[nd.children[i]].measure)
            }
            if (i < nd.n) {
                @Suppress("UNCHECKED_CAST")
                m = traits.combine(m, traits.measure(nd.keys[i] as K))
            }
        }
        nd.measure = m
    }

    /**
     * Split a 4-node child at position [childPos] of parent at [parentIdx].
     * The child has keys k0,k1,k2 and children c0,c1,c2,c3.
     * After split: left gets (k0; c0,c1), right gets (k2; c2,c3),
     * k1 is pushed into parent at childPos.
     */
    private fun splitChild(parentIdx: Int, childPos: Int) {
        val childIdx = nodes[parentIdx].children[childPos]

        // Save child data before allocNode() may cause issues
        val k0 = nodes[childIdx].keys[0]
        val k1 = nodes[childIdx].keys[1]
        val k2 = nodes[childIdx].keys[2]
        val c0 = nodes[childIdx].children[0]
        val c1 = nodes[childIdx].children[1]
        val c2 = nodes[childIdx].children[2]
        val c3 = nodes[childIdx].children[3]

        // Create right node with k2, c2, c3
        val rightIdx = allocNode()
        nodes[rightIdx].n = 1
        nodes[rightIdx].keys[0] = k2
        nodes[rightIdx].children[0] = c2
        nodes[rightIdx].children[1] = c3

        // Shrink child (left) to k0, c0, c1
        nodes[childIdx].n = 1
        nodes[childIdx].keys[0] = k0
        nodes[childIdx].keys[1] = null
        nodes[childIdx].keys[2] = null
        nodes[childIdx].children[0] = c0
        nodes[childIdx].children[1] = c1
        nodes[childIdx].children[2] = -1
        nodes[childIdx].children[3] = -1

        recomputeMeasure(childIdx)
        recomputeMeasure(rightIdx)

        // Insert mid_key (k1) into parent at childPos
        val parent = nodes[parentIdx]
        for (i in parent.n downTo childPos + 1) {
            parent.keys[i] = parent.keys[i - 1]
            parent.children[i + 1] = parent.children[i]
        }
        parent.keys[childPos] = k1
        parent.children[childPos + 1] = rightIdx
        parent.n++

        recomputeMeasure(parentIdx)
    }

    /**
     * Insert key into a non-full node's subtree.
     * Precondition: node at idx is not a 4-node.
     */
    private fun insertNonFull(idx: Int, key: K) {
        if (isLeaf(idx)) {
            val nd = nodes[idx]
            var pos = nd.n
            while (pos > 0 && traits.compare(key, nodeKey(idx, pos - 1)) < 0) {
                nd.keys[pos] = nd.keys[pos - 1]
                pos--
            }
            nd.keys[pos] = key
            nd.n++
            recomputeMeasure(idx)
            return
        }

        // Find child to descend into
        var pos = 0
        while (pos < nodes[idx].n && traits.compare(key, nodeKey(idx, pos)) >= 0) {
            pos++
        }

        // If that child is a 4-node, split it first
        if (is4Node(nodes[idx].children[pos])) {
            splitChild(idx, pos)
            if (traits.compare(key, nodeKey(idx, pos)) >= 0) {
                pos++
            }
        }

        insertNonFull(nodes[idx].children[pos], key)
        recomputeMeasure(idx)
    }

    fun insert(key: K) {
        if (root == -1) {
            root = allocNode()
            nodes[root].n = 1
            nodes[root].keys[0] = key
            recomputeMeasure(root)
            count++
            return
        }

        // If root is a 4-node, split it
        if (is4Node(root)) {
            val oldRoot = root
            root = allocNode()
            nodes[root].children[0] = oldRoot
            splitChild(root, 0)
        }

        insertNonFull(root, key)
        count++
    }

    fun clear() {
        nodes.clear()
        freeList.clear()
        root = -1
        count = 0
    }

    /** In-order traversal, calling [f] on each key. */
    fun forEach(f: (K) -> Unit) {
        forEachImpl(root, f)
    }

    @Suppress("UNCHECKED_CAST")
    private fun forEachImpl(idx: Int, f: (K) -> Unit) {
        if (idx == -1) return
        val nd = nodes[idx]
        for (i in 0..nd.n) {
            if (nd.children[i] != -1) {
                forEachImpl(nd.children[i], f)
            }
            if (i < nd.n) {
                f(nd.keys[i] as K)
            }
        }
    }

    /** Collect all keys in-order into a list. */
    fun collect(): MutableList<K> {
        val out = ArrayList<K>(count)
        forEach { out.add(it) }
        return out
    }

    private fun subtreeCount(idx: Int): Int {
        if (idx == -1) return 0
        val nd = nodes[idx]
        var c = nd.n
        for (i in 0..nd.n) {
            if (nd.children[i] != -1) c += subtreeCount(nd.children[i])
        }
        return c
    }

    /**
     * Find the element at which cumulative weight (as extracted by [weightOf])
     * reaches [target]. Returns a [WeightResult] with the key, the cumulative
     * weight before that key, the global index, and whether it was found.
     */
    fun findByWeight(target: Double, weightOf: (M) -> Double): WeightResult<K>? {
        if (root == -1) return null
        return findByWeightImpl(root, target, 0.0, 0, weightOf)
    }

    @Suppress("UNCHECKED_CAST")
    private fun findByWeightImpl(
        idx: Int, target: Double, cum: Double, globalIdx: Int,
        weightOf: (M) -> Double
    ): WeightResult<K>? {
        if (idx == -1) return null
        val nd = nodes[idx]
        var runningCum = cum
        var runningIdx = globalIdx

        for (i in 0..nd.n) {
            // Process child
            if (nd.children[i] != -1) {
                val childWeight = weightOf(nodes[nd.children[i]].measure)
                if (runningCum + childWeight >= target) {
                    return findByWeightImpl(nd.children[i], target, runningCum, runningIdx, weightOf)
                }
                runningCum += childWeight
                runningIdx += subtreeCount(nd.children[i])
            }

            if (i < nd.n) {
                val keyWeight = weightOf(traits.measure(nd.keys[i] as K))
                if (runningCum + keyWeight >= target) {
                    return WeightResult(nd.keys[i] as K, runningCum, runningIdx, true)
                }
                runningCum += keyWeight
                runningIdx++
            }
        }

        return null
    }

    /** Build a balanced tree from a sorted list. */
    fun buildFromSorted(sorted: List<K>) {
        clear()
        if (sorted.isEmpty()) return
        count = sorted.size
        root = buildRecursive(sorted, 0, sorted.size)
    }

    private fun buildRecursive(sorted: List<K>, lo: Int, hi: Int): Int {
        val n = hi - lo
        if (n <= 0) return -1

        if (n <= 3) {
            val idx = allocNode()
            nodes[idx].n = n
            for (i in 0 until n) nodes[idx].keys[i] = sorted[lo + i]
            recomputeMeasure(idx)
            return idx
        }

        if (n <= 7) {
            val mid = lo + n / 2
            val left = buildRecursive(sorted, lo, mid)
            val right = buildRecursive(sorted, mid + 1, hi)
            val idx = allocNode()
            nodes[idx].n = 1
            nodes[idx].keys[0] = sorted[mid]
            nodes[idx].children[0] = left
            nodes[idx].children[1] = right
            recomputeMeasure(idx)
            return idx
        }

        // For larger ranges, use 3-node
        val third = n / 3
        val m1 = lo + third
        val m2 = lo + 2 * third + 1
        val c0 = buildRecursive(sorted, lo, m1)
        val c1 = buildRecursive(sorted, m1 + 1, m2)
        val c2 = buildRecursive(sorted, m2 + 1, hi)
        val idx = allocNode()
        nodes[idx].n = 2
        nodes[idx].keys[0] = sorted[m1]
        nodes[idx].keys[1] = sorted[m2]
        nodes[idx].children[0] = c0
        nodes[idx].children[1] = c1
        nodes[idx].children[2] = c2
        recomputeMeasure(idx)
        return idx
    }
}
